File: leetcode.com/python/428_Serialize_and_Deserialize_N-ary_Tree.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""


# Idea: preorder recursive traversal; add number of children after root val, in order to know when to terminate.
class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: Node
        :rtype: str
        """
        nodeList = []
        self.serializeHelper(root, nodeList)
        return ','.join(map(str, nodeList))


    def serializeHelper(self, root, nodeList):
        if root is None:
            return
        nodeList.append(root.val)
        nodeList.append(len(root.children))
        for child in root.children:
            self.serializeHelper(child, nodeList)


    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: Node
        """
        if len(data) <= 0:
            return None
        nodeList = data.split(",")
        currentNodeIndexs = [0]
        deserializedData = self.deserializeHelper(nodeList, currentNodeIndexs)
        return deserializedData


    def deserializeHelper(self, nodeList, currentNodeIndexs):
        if currentNodeIndexs[0] == len(nodeList):
            return None
        root = Node(int(nodeList[currentNodeIndexs[0]]), [])
        currentNodeIndexs[0] += 1
        childrenSize = int(nodeList[currentNodeIndexs[0]])
        currentNodeIndexs[0] += 1
        for index in range(childrenSize):
            root.children.append(self.deserializeHelper(nodeList, currentNodeIndexs))
        return root


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/428_Serialize_and_Deserialize_N-ary_Tree.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var children: [Node]
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.children = []
 *     }
 * }
 */

class Codec {
    func serialize(_ root: Node?) -> String {
        var nodeList = [Int]()
        serializeDFSHelper(root, &nodeList)
        let encodedString = nodeList.reduce("") { $0 + "," + String($1) }
        return encodedString
    }
    
    func serializeDFSHelper(_ root: Node?, _ nodeList: inout [Int]) {
        guard let root = root else {
            return
        }
        nodeList.append(root.val)
        nodeList.append(root.children.count)
        for child in root.children {
            serializeDFSHelper(child, &nodeList)
        }
    }
    
    func deserialize(_ data: String) -> Node? {
        let nodeList = data.components(separatedBy: ",").flatMap { Int($0) }
        guard nodeList.count > 0 else {
            return nil
        }
        var currentIndex = 0
        return deserializeDFSHelper(nodeList, &currentIndex)
    }
    
    func deserializeDFSHelper(_ nodeList: [Int], _ currentIndex: inout Int) -> Node? {
        guard currentIndex + 1 < nodeList.count else {
            return nil
        }
        let node = Node(nodeList[currentIndex]), childrenCount = nodeList[currentIndex + 1]
        currentIndex += 2
        if childrenCount == 0 {
            return node
        }
        
        var children = [Node]()
        for i in 0..<childrenCount {
            let child = deserializeDFSHelper(nodeList, &currentIndex)
            children.append(child!)
        }
        node.children = children
        return node
    }
}

/**
 * Your Codec object will be instantiated and called as such:
 * let obj = Codec()
 * let ret_1: TreeNode? = obj.serialize(root)
 * let ret_2: Node? = obj.decode(data)
 */
